cloud-controller-manager/do/cloud.go

```diff
@@ -151,3 +158,105 @@ func (c *cloud) ScrubDNS(nameservers, searches []string) (nsOut, srchOut []strin
 func (c *cloud) HasClusterID() bool {
 	return false
 }
+
+type cloudResources interface {
+	DropletByID(ctx context.Context, id int) (*godo.Droplet, bool, error)
+	DropletByName(ctx context.Context, name string) (*godo.Droplet, bool, error)
+
+	Reload(ctx context.Context) error
+}
+
+type memResources struct {
+	dropletIDMap   map[int]*godo.Droplet
+	dropletNameMap map[string]*godo.Droplet
+	mutex          sync.RWMutex
+}
+
+func (c *memResources) Reload(ctx context.Context) error {
+	// Noop
+	return nil
+}
+
+func (c *memResources) DropletByID(ctx context.Context, id int) (droplet *godo.Droplet, found bool, err error) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	droplet, found = c.dropletIDMap[id]
+	return droplet, found, nil
+}
+
+func (c *memResources) DropletByName(ctx context.Context, name string) (droplet *godo.Droplet, found bool, err error) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	droplet, found = c.dropletNameMap[name]
+	return droplet, found, nil
+}
+
+type cachedAPI struct {
+	*memResources
+
+	client *godo.Client
+
+	ttl        time.Duration
+	expiration time.Time
+
+	now func() time.Time
+}
+
+func newCachedAPI(ttl time.Duration, client *godo.Client) *cachedAPI {
+	return &cachedAPI{
+		memResources: &memResources{},
+
+		client: client,
+		ttl:    ttl,
+		now:    time.Now,
+	}
+}
+
+func (c *cachedAPI) Reload(ctx context.Context) error {
```

Comment:
I see a few concerns with the way `Reload` needs to be integrated and how it is implemented:

- due to the lock hold, `Reload` will block concurrent callers until the (potentially long) operation completes. This could lead to delays and maybe even timeouts on the caller end.
- The cost of reloading are amortized across all calls, which means that some consumers will pay more than others. This may lead to quality degradation scenarios that are harder to identify.
- A failed reload causes the entire operation to fail even in the case where we may have cached (albeit somewhat stale) droplet data that we could return.
- We have to build in logic that determines whether it's time to reload and calculate the future reload point in time.
- We have to make sure to invoke the `Reload` side-effect on every method that wants to look up droplets, which is something that could be forgotten if we ever add more methods or refactor the code.

Some of these shortcomings can likely be avoided by improving the current implementation. However, I want to propose an alternative design that may be better suited systematically: let's outsource droplet management into a dedicated goroutine responsible for reloading and serving lookups. The key benefits would be that it could run at a static periodicity, return droplet data via a (inherently synchronized) channel, require an (internal) lock only when swapping out old data via new one, decouple reload from lookup cost, and keep returning stale data until the reload completes.

WDYT?

Reply:
Agree with many of your concerns. I steered away from doing a periodic reload on a separate goroutine for 2 reasons:

1. This could just be a conceptual difference in how we are viewing a sync loop in a separate goroutine. Both approaches achieve the goal of invalidating/refreshing a cache, but conceptually the sync loop to me feels like we are trying to sync state between the DigitalOcean api and ccm. I feel like the cloud provider interface lends itself better to on demand querying a cloud's non-kubernetes api (the caching in the current approach only serving as a short-circuit if fresh enough). Again, I feel like this is just a conceptual difference, not really a material one.

2. There is rate limit at play on the public api token that ccm uses, so minimizing the number of requests made would be ideal. A sync loop that continually runs would use up a portion of the alloted requests. Not a big deal, just something that came to mind.

Like you said, there are quite a few things we can do to improve the current implementation to address some of your concerns, but I'm happy to swtich over to a sync loop for managing the cache as well. Let me know if you would still prefer that.

Reply:
With regards to 1: The cloud provider interface includes a method `Initialize` for the dedicated purpose of initializing asynchronous activities. [From the in-code comment](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/68f5e06565659dcd45c3c608dd05d542ad111e66/vendor/k8s.io/kubernetes/pkg/cloudprovider/cloud.go#L33-L35):

>	// Initialize provides the cloud with a kubernetes client builder and may spawn goroutines
	// to perform housekeeping activities within the cloud provider.

So I believe it'd be okay to implement a cloud/CCM synchronization for performance reasons from a conceptual point of view.

Fair point about the rate limiting aspect, I missed to take that into account. Looks like the DigitalOcean API currently [permits 5,000 requests per hour and token](https://developers.digitalocean.com/documentation/v2/#rate-limit). With a static update frequency of once per minute (as currently configured), that's 60 extra hourly requests or 0.83% of the overall quota. I can see us possibly doubling the frequency to twice per minute, so overall that'd be a good precent overhead.

Whether that's acceptable or not likely depends on a user's API usage pattern. What I'd think is that clusters which require many updates would also frequently reload with the current implementation anyway while users with infrequent reloads may not be bothered much by the few extra reloads. Users sharing tokens across many (high-volume) clusters are likely better off using dedicated API tokens per cluster for reasons not only related to CCM. Admittedly, that's a fair amount of assumptions and "you should do this and that" advice.

So while I'd lean towards adapting the goroutine approach allowing us to address the concerns I outlined above now and dealing with potential quota issues if/when they arise, I'm also okay with sticking to the PR approach. The only bits I'd then like to ask for is to use cached droplet data if the reload fails (and [emit an event](https://github.com/kubernetes/sample-controller/blob/0e5b089d858f35daf461613074207039f23cac77/controller.go#L291) to inform the user [point no. 3 from my list above]) and maybe also _try_ to address the concurrency / blocking user issue. (I say try because I can see this may not be trivial to implement with the current approach, so it's not a must-have) [point no. 1].

I'll let you make the call. Once decided, we can proceed with the code review.

Thanks!